### PR TITLE
NET-89:  TURN reactive Implementation

### PR DESCRIPTION
**Pull request by abhishek9686**
## Describe your changes
 -> Peer connections are checked periodically for connection status
 -> Switches to using turn if the connection is not healthy
## Provide link to Netmaker PR if required
https://github.com/gravitl/netmaker/pull/2290
## Provide testing steps
- [ ] Test on devices which are hard to reach, the connection should be automatically switched to turn after a min if there is no handshake, or if the handshake is older than 3 minutes
- [ ]  On running pull cmd, the connection should be reset to the actual endpoint of peers if they were using turn

## Checklist before requesting a review
- [x] My changes affect only 10 files or less.
- [x] I have performed a self-review of my code and tested it.
- [x] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [x] My functions are <= 80 lines.
- [ ] I have had my code reviewed by a peer.
- [ ] My unit tests pass locally.
- [x] Netclient & Netmaker are awesome.

### CODE REVIEW

File: nmproxy/turn/turn.go [hunk 1]
```diff
@@ -15,10 +15,12 @@ import (
 	"github.com/devilcove/httpclient"
 	"github.com/gravitl/netclient/auth"
 	ncconfig "github.com/gravitl/netclient/config"
+	"github.com/gravitl/netclient/ncutils"
 	"github.com/gravitl/netclient/nmproxy/config"
 	"github.com/gravitl/netclient/nmproxy/models"
 	"github.com/gravitl/netclient/nmproxy/packet"
 	"github.com/gravitl/netclient/nmproxy/server"
+	wireguard "github.com/gravitl/netclient/nmproxy/wg"
```
Comment (mattkasun) [inline nmproxy/turn/turn.go:23]: why not name the package wireguard if you are going to import it this way

File: nmproxy/nm-proxy.go [hunk 2]
```diff
@@ -46,14 +46,21 @@ func Start(ctx context.Context, wg *sync.WaitGroup,
 	if err != nil {
 		logger.FatalLog("failed to create proxy: ", err.Error())
 	}
+	proxyWaitG := &sync.WaitGroup{}
```
Comment (mattkasun) [inline nmproxy/nm-proxy.go:49]: add a comment as to why the extra waitgroup is required.

File: nmproxy/turn/turn_helper.go [hunk 3]
```diff
@@ -16,87 +19,201 @@ import (
...
-						Server:            signal.Server,
-						FromHostPubKey:    signal.ToHostPubKey,
-						TurnRelayEndpoint: hostTurnCfg.TurnConn.LocalAddr().String(),
-						ToHostPubKey:      signal.FromHostPubKey,
-						Reply:             true,
-					})
-					hostTurnCfg.Mutex.RUnlock()
+		}
+
+		// if respone to the signal you sent,then don't signal back
```
Comment (mattkasun) [inline nmproxy/turn/turn_helper.go:97]: typo  respone-->response

**Review by mattkasun — CHANGES_REQUESTED** [review]

File: nmproxy/nm-proxy.go (hunk 2 above)
Comment (abhishek9686) [reply nmproxy/nm-proxy.go:49]: we have cleanups happening in the `nmproxy.Start()` on shutdown, which should  only happen after all the threads in the proxy are closed and to do this parent wait group cannot be used to wait

**Review by mattkasun — APPROVED** [review]

**Pull request merged**